File: CS/WindowsApplication3/Main.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraBars.Docking;
using DevExpress.Utils.Drawing;
using DevExpress.XtraBars.Docking.Paint;
using DevExpress.XtraBars.Docking.Helpers;
using DevExpress.Utils;
using DevExpress.XtraBars.Styles;


namespace DXSample {
    public partial class Main: XtraForm {
        public Main() {
            InitializeComponent();
        }

        private void OnFormLoad(object sender, EventArgs e) {
            barAndDockingController1.PaintStyles.Add(new MyPaintStyle(barAndDockingController1.PaintStyles));
            barAndDockingController1.PaintStyleName = "FullTabTextStyle";
        }
    }

    public class MyPaintStyle : SkinBarManagerPaintStyle
    {
        public MyPaintStyle(BarManagerPaintStyleCollection collection)
            : base(collection)
        {
        }
        public override string Name { get { return "FullTabTextStyle"; } }
        protected override void CreatePainters()
        {
            base.CreatePainters();
            fElementsPainter = new MyDockElementsSkinPainter(this);
        }
    }

    public class MyDockElementsSkinPainter : DockElementsSkinPainter
    {
        public MyDockElementsSkinPainter(SkinBarManagerPaintStyle paintStyle) : base(paintStyle) { }
        protected override void CreateElementPainters()
        {
            base.CreateElementPainters();
            this.fHideBarPainter = new MyHideBarPainter(this);
        }
    }

    public class MyHideBarPainter : HideBarSkinPainter
    {
        DockLayout layout = null;
        public MyHideBarPainter(DockElementsSkinPainter painter) : base(painter) { }
        public override int GetButtonWidth(DockLayout dockLayout, AppearanceObject appearance, bool isHorizontal)
        {
            int result = 0;
            Graphics g = Painter.AddGraphics(null);
            if (dockLayout.Tabbed)
            {
                try
                {
                    int imageSize = (isHorizontal ? GetImageSize(dockLayout).Width : GetImageSize(dockLayout).Height);
                    int defImageSize = (isHorizontal ? GetDefaultImageSize(dockLayout).Width : GetDefaultImageSize(dockLayout).Height);
                    int imageInterval = (imageSize > 0 ? HideBarHorzInterval : 0);
                    for (int i = 0; i < dockLayout.Count; i++)
                    {
                        if (dockLayout[i].Visibility == DockVisibility.Visible)
                        {
                            layout = dockLayout[i];
                            int width = GetTabButtonWidth(dockLayout[i]);
                            result += width + TabButtonsInterval;
                        }
                    }
                }
                finally
                {
                    Painter.ReleaseGraphics();
                }
            }
            else return base.GetButtonWidth(dockLayout, appearance, isHorizontal);
            return result;
        }
       
        public override void DrawHideBarButton(DrawHideBarButtonArgs args)
        {
            Rectangle bounds = args.Bounds;
            Rectangle tabRect = bounds, activeTabBounds = bounds;
            DockLayout dockLayout = args.DockLayout;
            if (args.DockLayout.Tabbed)
            {
                for (int i = 0; i < dockLayout.Count; i++)
                {
                    int size = GetTabButtonWidth(dockLayout[i]);
                    if (args.IsVertical)
                    {
                        if (dockLayout.ActiveChildIndex != i)
                            DrawInactiveTabContainerItem(args, dockLayout[i], GetTabRectByIndex(i, dockLayout, bounds, args.IsVertical), i == 0, false);
                    }
                    else
                    {
                        if (dockLayout.ActiveChildIndex != i)
                            DrawInactiveTabContainerItem(args, dockLayout[i], GetTabRectByIndex(i, dockLayout, bounds, args.IsVertical), i == 0, false);
                    }
                }
                activeTabBounds = GetTabRectByIndex(args.DockLayout.ActiveChildIndex, dockLayout, bounds, args.IsVertical);
                dockLayout = args.DockLayout.ActiveChild;
                bounds = activeTabBounds;
                args = new DrawHideBarButtonArgs(args, activeTabBounds, args.IsActive, ObjectState.Normal);
            }
            DrawHideBarButtonEdges(args);
            if (dockLayout == null) return;
            DrawHideBarControlButtonContent(args, bounds, dockLayout, null, EdgePositions.None);
        }

        protected override Rectangle GetTabChildButtonBounds(DockLayout dockLayout, int index, Rectangle bounds, bool isVertical)
        {
            return GetTabRectByIndex(index, dockLayout, bounds, isVertical);
        }

        Rectangle GetTabRectByIndex(int index, DockLayout dockLayout, Rectangle bounds, bool isVertical)
        {
            Rectangle tabRect = bounds;
            for (int i = 0; i < dockLayout.Count; i++)
            {
                int size = GetTabButtonWidth(dockLayout[i]);
                if (isVertical)
                {
                    tabRect = new Rectangle(tabRect.X, tabRect.Y, tabRect.Width, size);
                    if (i == index) return tabRect;
                    tabRect.Y += size + TabButtonsInterval;
                }
                else
                {
                    tabRect = new Rectangle(tabRect.X, tabRect.Y, size, tabRect.Height);
                    if (i == index) return tabRect;
                    tabRect.X += size + TabButtonsInterval;
                }
            }
            return tabRect;
        }

        protected int GetTabButtonWidth(DockLayout layout)
        {
            AppearanceObject appearance = layout.DockManager.ActivePanel == layout.Panel ? layout.ActiveTabAppearance : layout.TabsAppearance;
            Graphics g = Painter.AddGraphics(null);
            int imageSize = (layout.IsHorizontal ? GetImageSize(layout).Width : GetImageSize(layout).Height);
            int defImageSize = (layout.IsHorizontal ? GetDefaultImageSize(layout).Width : GetDefaultImageSize(layout).Height);
            int imageInterval = (imageSize > 0 ? HideBarHorzInterval : 0);
            int width = Math.Max(DockElementsPainter.CalcTextSize(g, appearance, layout.TabText).Width + imageSize + imageInterval, defImageSize) + 10;
            Painter.ReleaseGraphics();
            return width;
        }
    }
}

File: VB/WindowsApplication3/Main.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraBars.Docking
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraBars.Docking.Paint
Imports DevExpress.XtraBars.Docking.Helpers
Imports DevExpress.Utils
Imports DevExpress.XtraBars.Styles


Namespace DXSample
	Partial Public Class Main
		Inherits XtraForm
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub OnFormLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			barAndDockingController1.PaintStyles.Add(New MyPaintStyle(barAndDockingController1.PaintStyles))
			barAndDockingController1.PaintStyleName = "FullTabTextStyle"
		End Sub
	End Class

	Public Class MyPaintStyle
		Inherits SkinBarManagerPaintStyle
		Public Sub New(ByVal collection As BarManagerPaintStyleCollection)
			MyBase.New(collection)
		End Sub
		Public Overrides ReadOnly Property Name() As String
			Get
				Return "FullTabTextStyle"
			End Get
		End Property
		Protected Overrides Sub CreatePainters()
			MyBase.CreatePainters()
			fElementsPainter = New MyDockElementsSkinPainter(Me)
		End Sub
	End Class

	Public Class MyDockElementsSkinPainter
		Inherits DockElementsSkinPainter
		Public Sub New(ByVal paintStyle As SkinBarManagerPaintStyle)
			MyBase.New(paintStyle)
		End Sub
		Protected Overrides Sub CreateElementPainters()
			MyBase.CreateElementPainters()
			Me.fHideBarPainter = New MyHideBarPainter(Me)
		End Sub
	End Class

	Public Class MyHideBarPainter
		Inherits HideBarSkinPainter
		Private layout As DockLayout = Nothing
		Public Sub New(ByVal painter As DockElementsSkinPainter)
			MyBase.New(painter)
		End Sub
		Public Overrides Function GetButtonWidth(ByVal dockLayout As DockLayout, ByVal appearance As AppearanceObject, ByVal isHorizontal As Boolean) As Integer
			Dim result As Integer = 0
			Dim g As Graphics = Painter.AddGraphics(Nothing)
			If dockLayout.Tabbed Then
				Try
					Dim imageSize As Integer
					If isHorizontal Then
						imageSize = (GetImageSize(dockLayout).Width)
					Else
						imageSize = (GetImageSize(dockLayout).Height)
					End If
					Dim defImageSize As Integer
					If isHorizontal Then
						defImageSize = (GetDefaultImageSize(dockLayout).Width)
					Else
						defImageSize = (GetDefaultImageSize(dockLayout).Height)
					End If
					Dim imageInterval As Integer
					If imageSize > 0 Then
						imageInterval = (HideBarHorzInterval)
					Else
						imageInterval = (0)
					End If
					For i As Integer = 0 To dockLayout.Count - 1
						If dockLayout(i).Visibility = DockVisibility.Visible Then
							layout = dockLayout(i)
							Dim width As Integer = GetTabButtonWidth(dockLayout(i))
							result += width + TabButtonsInterval
						End If
					Next i
				Finally
					Painter.ReleaseGraphics()
				End Try
			Else
				Return MyBase.GetButtonWidth(dockLayout, appearance, isHorizontal)
			End If
			Return result
		End Function

		Public Overrides Sub DrawHideBarButton(ByVal args As DrawHideBarButtonArgs)
			Dim bounds As Rectangle = args.Bounds
			Dim tabRect As Rectangle = bounds, activeTabBounds As Rectangle = bounds
			Dim dockLayout As DockLayout = args.DockLayout
			If args.DockLayout.Tabbed Then
				For i As Integer = 0 To dockLayout.Count - 1
					Dim size As Integer = GetTabButtonWidth(dockLayout(i))
					If args.IsVertical Then
						If dockLayout.ActiveChildIndex <> i Then
							DrawInactiveTabContainerItem(args, dockLayout(i), GetTabRectByIndex(i, dockLayout, bounds, args.IsVertical), i = 0, False)
						End If
					Else
						If dockLayout.ActiveChildIndex <> i Then
							DrawInactiveTabContainerItem(args, dockLayout(i), GetTabRectByIndex(i, dockLayout, bounds, args.IsVertical), i = 0, False)
						End If
					End If
				Next i
				activeTabBounds = GetTabRectByIndex(args.DockLayout.ActiveChildIndex, dockLayout, bounds, args.IsVertical)
				dockLayout = args.DockLayout.ActiveChild
				bounds = activeTabBounds
				args = New DrawHideBarButtonArgs(args, activeTabBounds, args.IsActive, ObjectState.Normal)
			End If
			DrawHideBarButtonEdges(args)
			If dockLayout Is Nothing Then
				Return
			End If
			DrawHideBarControlButtonContent(args, bounds, dockLayout, Nothing, EdgePositions.None)
		End Sub

		Protected Overrides Function GetTabChildButtonBounds(ByVal dockLayout As DockLayout, ByVal index As Integer, ByVal bounds As Rectangle, ByVal isVertical As Boolean) As Rectangle
			Return GetTabRectByIndex(index, dockLayout, bounds, isVertical)
		End Function

		Private Function GetTabRectByIndex(ByVal index As Integer, ByVal dockLayout As DockLayout, ByVal bounds As Rectangle, ByVal isVertical As Boolean) As Rectangle
			Dim tabRect As Rectangle = bounds
			For i As Integer = 0 To dockLayout.Count - 1
				Dim size As Integer = GetTabButtonWidth(dockLayout(i))
				If isVertical Then
					tabRect = New Rectangle(tabRect.X, tabRect.Y, tabRect.Width, size)
					If i = index Then
						Return tabRect
					End If
					tabRect.Y += size + TabButtonsInterval
				Else
					tabRect = New Rectangle(tabRect.X, tabRect.Y, size, tabRect.Height)
					If i = index Then
						Return tabRect
					End If
					tabRect.X += size + TabButtonsInterval
				End If
			Next i
			Return tabRect
		End Function

		Protected Function GetTabButtonWidth(ByVal layout As DockLayout) As Integer
			Dim appearance As AppearanceObject
			If layout.DockManager.ActivePanel Is layout.Panel Then
				appearance = layout.ActiveTabAppearance
			Else
				appearance = layout.TabsAppearance
			End If
			Dim g As Graphics = Painter.AddGraphics(Nothing)
			Dim imageSize As Integer
			If layout.IsHorizontal Then
				imageSize = (GetImageSize(layout).Width)
			Else
				imageSize = (GetImageSize(layout).Height)
			End If
			Dim defImageSize As Integer
			If layout.IsHorizontal Then
				defImageSize = (GetDefaultImageSize(layout).Width)
			Else
				defImageSize = (GetDefaultImageSize(layout).Height)
			End If
			Dim imageInterval As Integer
			If imageSize > 0 Then
				imageInterval = (HideBarHorzInterval)
			Else
				imageInterval = (0)
			End If
			Dim width As Integer = Math.Max(DockElementsPainter.CalcTextSize(g, appearance, layout.TabText).Width + imageSize + imageInterval, defImageSize) + 10
			Painter.ReleaseGraphics()
			Return width
		End Function
	End Class
End Namespace
